Used a key prefix to enable immediate cache resets

allouis opened the pull request:
## Why

The previous `reset()` scanned every matching key and DEL'd them one by one. It was O(n), slow against large caches, and needed cluster-aware routing to hit the right primary node. The slowness added load to the Redis cluster, and it meant that we had to limit the number of sites we could configure to use the cache.

This PR swaps SCAN-and-DEL for prefix rotation. Every key is stored under a random hash that's prepended at write time, so rotating the hash makes every entry unreachable in a single SET. Old entries age out via Redis's own eviction and TTL.

## Changes

- `prefix_hash` is lazily initialized on first read. Concurrent callers within a process share an in-flight promise; `SET NX` collapses the cross-process race so losers read back the winning value.
- `prefix_hash` reads and writes bypass cache-manager and go through the raw redis client. The meta-key holds a raw hex string, not JSON, so cache-manager's wrapper would choke on it; for reset we also want an unconditional overwrite.
- The SCAN-based helpers (`#getKeys`, `#scanNodeForKeys`, `#getPrimaryRedisNode`) and `_removeKeyPrefix` are deleted. They only existed to support the old SCAN-based `reset()`.
- The `cluster-key-slot` dependency is no longer used.

## Details

- `prefix_hash` is written without a TTL, and our Redis cluster uses `volatile-lru` eviction policy, so these keys won't be evicted.
- Every `get` and `set` now costs an extra round-trip to fetch the current `prefix_hash`. For cache hits that doubles the latency. We get O(1) resets in return, and we're accepting the cost for now. We are going to monitor Redis performance as we add new sites, and we might revisit this if it's an issue.


## Testing

The 12 integration tests from #27467 (already in main) continue to pass; they exercise the invalidation contract against real Redis.

allouis committed:
Used a key prefix to enable immediate cache resets

ref https://linear.app/ghost/issue/HKG-1708

The previous reset() scanned every matching key and DEL'd them one by
one. It was O(n), slow against large caches, and needed cluster-aware
routing to hit the right primary node.

This replaces SCAN-and-DEL with prefix rotation. Every key is stored
under a random hash that's prepended at write time, so rotating the
hash makes every entry unreachable in one SET. Old entries age out via
Redis's own eviction and TTL.

The SCAN-based helpers (#getKeys, #scanNodeForKeys,
#getPrimaryRedisNode) and _removeKeyPrefix go too. They only existed
to support the old SCAN-based reset() and the now-stubbed keys().
The cluster-key-slot dependency is no longer needed.

To avoid races on first init, concurrent callers within a process
share a single in-flight promise, and SET NX collapses the
cross-process race so only one writer wins. reset() keeps the old
try/catch around redis errors so they get logged and don't bubble up.

Two trade-offs worth flagging:

prefix_hash is written without a TTL, but redis can still evict it
under allkeys-lru / allkeys-random policies. Eviction silently wipes
the cache. Operators who need strict invalidation guarantees should
use noeviction.

Every cache op now does an extra GET to fetch the current hash. For
cache hits that doubles the round-trip count. Tracked via the new
cache-prefix-hash-fetch metric.

allouis reviewed (commented):
This reads the `prefix_hash` entry from Redis on every read - which may not be desired.

The thinking is that we can do an immediate reset on our side by changing/deleting that key if wanted. 

We can also look at caching it in memory, but that is something we want to get away from...

allouis commented on ghost/adapter-cache-redis/lib/AdapterCacheRedis.js:
```diff
@@ -219,28 +241,18 @@ class AdapterCacheRedis extends BaseCacheAdapter {
 
+    async cyclePrefixHash() {
+        const value = crypto.randomBytes(12).toString('hex');
+        await this.cache.set(this._keyPrefix + 'prefix_hash', value, {ttl: 0});
```
Setting the TTL to 0 means that the key should not expire in Redis

allouis commented on ghost/core/core/server/adapters/lib/redis/AdapterCacheRedis.js:
```diff
@@ -119,11 +139,12 @@ class AdapterCacheRedis extends BaseCacheAdapter {
-            return `${this.keyPrefix}${key}`;
+    async _buildKey(key) {
+        const keyPrefix = await this.keyPrefix();
+        if (keyPrefix) {
```
This is always truthy now, can remove the check and alternate branch

allouis commented:
@coderabbitai review

kevinansfield reviewed (commented):
`_get` wraps the Redis fetch in a timeout but this introduces an extra fetch before `_get` to read the key from Redis. That key fetch isn't wrapped in the fetch timeout, should it be? I think the outcome is that the cache being down/slow can potentially cause fallback to the real API to take longer than expected

allouis committed:
Bounded the whole get() under getTimeoutMilliseconds

ref https://linear.app/ghost/issue/HKG-1708

The previous patch wrapped only the data fetch (cache.get) under
getTimeoutMilliseconds. After the prefix-rotation change the prefix
fetch became part of every read, but it was unbounded - so a slow or
stalled redis would let cache.get() hang past the configured timeout.

Lift the timeout out of _get and apply it to the buildKey + cache.get
sequence as a single budget. On timeout we resolve as a MISS so the
existing fallback path runs (fetchData + set if provided, otherwise
return null). _get goes away; it was only the timeout wrapper.

Adds an integration test that slows the prefix_hash fetch (separate
client) and asserts the same timeout behaviour. Renames the existing
slow-fetch test to clarify it covers the data path.

allouis commented:
@kevinansfield Thanks for looking! You're right, I've added an integration test to catch the key fetching timeouts, and I've updated `get` so that the timeout wraps both the key fetch and the value fetch!

vershwal commented on ghost/core/core/server/adapters/lib/redis/AdapterCacheRedis.js:
```diff
@@ -237,7 +241,7 @@ class AdapterCacheRedis extends BaseCacheAdapter {
      */
     async set(key, value) {
-        const internalKey = this._buildKey(key);
+        const internalKey = await this._buildKey(key);
```
We can move `_buildKey` inside the try-catch block. `_buildKey` is now async and hits Redis (via `prefixHash()` → `redisClient.get`), so it can throw. Here the `_buildKey` call sits above the try-catch, so a Redis failure would propagate uncaught to the caller.

vershwal commented on ghost/core/core/server/adapters/lib/redis/AdapterCacheRedis.js:
```diff
@@ -164,27 +162,34 @@ class AdapterCacheRedis extends BaseCacheAdapter {
+                debug('get', key, 'timeout');
+                resolve({internalKey: null, result: null});
+            }, this.getTimeoutMilliseconds);
+            lookup.then((value) => {
```
We could add a small `.catch()` here as well. Since lookup is async, if Redis fails while fetching the key or value, that rejection would currently go unhandled. The timeout still makes this behave like a cache miss, so it’s not a functional issue, but handling the rejection would avoid unnecessary `unhandledRejection` in the logs

vershwal reviewed (commented):
Overall the PR looks good. I’ve added a couple of comments, but nothing major. Let me know what you think.

allouis committed:
Caught redis errors from the now-async _buildKey

ref https://linear.app/ghost/issue/HKG-1708

_buildKey became async with the prefix-rotation work, which means it
now hits redis and can throw. Two callsites needed to catch the new
failure mode:

set() built the internal key outside its try/catch, so a redis
failure on the prefix fetch escaped to the caller. Moved the
_buildKey call inside the try.

#lookupWithTimeout had no rejection handler on the lookup promise,
so a failed buildKey or cache.get would dangle as an unhandled
rejection until the timeout fired and treated it as a MISS. Added a
catch handler that resolves the same way the timeout does, which
keeps the outer behaviour identical and silences the dangling
rejection.

daniellockyer commented:
😍
